_posts/2024-03-20-kraft-migration.md

```diff
@@ -0,0 +1,201 @@
+---
+layout: post
+title:  "From ZooKeeper to KRaft: how the Kafka migration works"
+date: 2024-03-20
+author: paolo_patierno
+---
+
+The Apache Kafka project has been using [Apache ZooKeeper](https://zookeeper.apache.org/) since its inception to store metadata.
+Registered brokers, the current controller and topics' configuration are just a few of the data that ZooKeeper stores for supporting a running Kafka cluster.
+Using a dedicated and centralized system for maintaining cluster metadata and offload leader election was the right approach at that time, because ZooKeeper was battle tested and it is purpose built for providing distributed coordination.
+But taking care of an additional component alongside your Kafka cluster is not that simple.
+Furthermore, ZooKeeper has become a bottleneck that limits the amount of partitions that a single broker can handle.
+For these and other reasons, the Kafka community opened the "famous" [KIP-500](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum) in late 2019 and started to work on the ZooKeeper removal.
+
+<!--more-->
+
+### What is "wrong" with ZooKeeper?
+
+As already mentioned, Kafka has been using ZooKeeper as its metadata management system for several functions:
+
+* **Cluster membership**: each broker, joining the Kafka cluster, registers itself with a ephemeral znode on ZooKeeper;
+* **Controller election**: when a broker starts, it tries to take the "controller" role by creating the ephemeral `/controller` znode on ZooKeeper. If that znode already exists, it indicates which broker is the current controller;
+* **Topics configuration**: all the topics' configuration parameters are stored in ZooKeeper together with information like the number of partitions and the current broker assignments for the replicas;
+* **Access Control Lists (ACLs)**: when a client connects to the cluster, it could be authenticated and authorized to read or write on several topics based on the ACLs stored in ZooKeeper by the built-in Authorizer;
+* **Quotas**: brokers can limit resources used by clients in terms of network bandwidth and CPU utilization via quotas stored in ZooKeeper by the built-in Quotas provider;
+
+By using the `zookeeper-shell` tool, it is possible to connect to a ZooKeeper ensemble and see all the znodes.
+
+```shell
+bin/zookeeper-shell.sh <zookeeper-ip-address>:2181
```

Comment:
There seems to be missing context how/where to use it which I guess suggests this works with Strimzi. But obviously, in Strimzi you cannot do this. I wonder if you should use `kubectl exec` and port `12181` for this so that users can try it?

Reply:
That's not on Strimzi. This is about migration in general so there is no assumption you are running on Strimzi. It could be any bare-metal or vms environment so you run it from where you can reach a ZooKeeper node assuming port 2181 is open. Tbh I didn't have the command at the beginning to not be so specific and added it after a suggestion by @mimaison.

Reply:
No there are no Strimzi examples anymore :-)
I split the article, so the Strimzi part is now a new PR https://github.com/strimzi/strimzi.github.io/pull/424 for the next blog post. This one is really generic around ZooKeeper to KRaft migration wherever you are running the system.

Reply:
If you are referring to the usage of `/var/lib/kafka/data-0/kafka-log0` for the log dirs, yes it's coming from Strimzi which was how I was running the cluster. But I am specifying that the path represents the log dirs. I could also change it in something different like just `/tmp/kafka-log` or other. But then, I don't think there are other references to Strimzi.

Reply:
The example right below that is clearly Strimzi. So you should use another one if you want to make it Stirmzi-independent. You should also make it clear in the text then that it is what you would see any some random KAfka cluster and not something you can do in Strimzi.

Reply:
Right! The endpoints! I don't want to make stuff much more complicated, I will mention I was using a Strimzi-operated cluster to get the examples and using the right commands.

Reply:
It is on a Strimzi blog. So I think if the command works on Strimzi out of the box, you do not need to explain much more 🤷

Reply:
I fixed the commands to run and the output adding some context about running them in a Strimzi-operated cluster.